gnu/usr.bin/avahi/man/avahi-publish.1
.TH avahi-publish-service 1 User Manuals
.SH NAME
avahi-publish-service \- Register an mDNS/DNS-SD service or host name or address mapping using the Avahi daemon
.SH SYNOPSIS
\fBavahi-publish -s [\fIoptions\fB] \fIname\fB \fIservice-type\fB \fIport\fB [\fITXT data\fB ...]

avahi-publish-service [\fIoptions\fB] \fIname\fB \fIservice-type\fB \fIport\fB [\fITXT data\fB ...]

avahi-publish -a [\fIoptions\fB] \fIhost name\fB \fIaddress\fB

avahi-publish-address [\fIoptions\fB] \fIhost name\fB \fIaddress\fB
\f1
.SH DESCRIPTION
Register an mDNS/DNS-SD service or host name/address mapping using the Avahi daemon.
.SH OPTIONS
When calling in service registration mode, specify a DNS-SD service name (e.g. "Lennart's Files"), a service type (e.g. _http._tcp) and an IP port number for the service, optionally followed by any number of TXT record strings on the command line. When calling in address/host name registration mode specify a fully qualified host name and an address (IPv4 or IPv6).
.TP
\fB-s | --service\f1
Register a service. avahi-publish-service is equivalent to avahi-publish -s.
.TP
\fB-a | --address\f1
Register an address/host name mapping. avahi-publish-address is equivalent to avahi-publish -a.
.TP
\fB-v | --verbose\f1
Enable verbose mode.
.TP
\fB-H | --host=\f1 \fIHOSTNAME\f1
Specify a host name for this service, in case it doesn't reside on the local host. This host name needs to be fully qualified and resolvable using mDNS or unicast DNS.
.TP
\fB-d | --domain=\f1 \fIDOMAIN\f1
Publish the service in the specified domain. If omitted the Avahi daemon will publish it in its default domain (usually .local).
.TP
\fB--subtype=\f1 \fISUBTYPE\f1
Register the service with an additional subtype in addition to the main type. DNS-SD subtypes have the form _anon._sub._ftp._tcp, where _anon is the identifier of the subtype and _ftp._tcp is the main type. You may pass this option multiple times to register the service with multiple subtypes.
.TP
\fB-f | --no-fail\f1
Don't fail if the daemon is not found running. Instead, wait until it appears. If it disconnects, try to reconnect.
.TP
\fB-h | --help\f1
Show help
.TP
\fB-V | --version\f1
Show version information.
.SH AUTHORS
The Avahi Developers <avahi (at) lists (dot) freedesktop (dot) org>; Avahi is available from \fBhttp://avahi.org/\f1
.SH SEE ALSO
\fBavahi-resolve(1)\f1, \fBavahi-browse(1)\f1, \fBavahi-daemon(8)\f1
.SH COMMENTS
This man page was written using \fBxml2man(1)\f1 by Oliver Kurth.


libs-gui/NEWS
1 NEWS
******

The currently released version of the library is '0.28.0'.

1.1 Noteworthy changes in version '0.28.0'
==========================================

This version adds support for modern XIB files and many new classes.
Plus the usual bunch of bug fixes.

   * Support loading of document XIB files.
   * Improve Key Value Binding for NSArrayController and add more
     bindings.
   * Better support for multi monitor usage and other improvement in the
     backend integration.
   * Add classes NSFontCollection, NSColorSampler, NSSpeechRecognizer,
     NSAppearance, NSPDFInfo, NSPICTImageRep, NSCIImageRep,
     NSPDFImageRep, NSPDFPanel, NSDataAsset, NSDatePicker,
     NSDatePickerCell, NSPredicateEditor, NSPredicateEditorRowTemplate,
     NSRuleEditor, NSGestureRecognizer, NSButtonTouchBarItem,
     NSCandidateListTouchBarItem, NSClickGestureRecognizer,
     NSColorPickerTouchBarItem, NSCustomTouchBarItem,
     NSGroupTouchBarItem, NSMagnificationGestureRecognizer,
     NSPanGestureRecognizer, NSPickerTouchBarItem,
     NSPopoverTouchBarItem, NSPressGestureRecognizer,
     NSRotationGestureRecognizer, NSSharingServicePickerTouchBarItem,
     NSSliderTouchBarItem, NSStepperTouchBarItem, NSTouchBarItem,
     NSTouchBar, NSTouch, NSDockTile.
   * Implement NSEPSImageRep.
   * Better encoding handling in RTF files.
   * Theming and drawing improvements.
   * Increase small font size to 10.
   * New cursor and stepper images.
   * Move NSFileWrapper to Foundation.
   * Fixed build on Debian GNU/kFreeBSD.
   * With command line argument -autolaunch YES, do not activate the
     application when -activateIgnoringOtherApps: is invoked.
   * Improvements to WindowMaker compatibility (e.g.  WMFHideApplication
     support).
   * Lowered NSFloatingWindowLevel by one to distinguish floating panels
     from menus.

1.2 Noteworthy changes in version '0.27.0'
==========================================

This version includes numerous bugfixes, compatibility improvements and
other changes accumulated over the last year.  It also enables work to
be done on integrating NSViews with a Core Animation renderer.

   * Make targetForAction safer.
   * Speed up menu updates.
   * Clean up speech tool compilation and switch to newer interface.
   * Fix bug in CUPS subclassing introduced in last release.
   * Minor improvements to typesetting.
   * Add NSIsControllerMarker.
   * Fix tracking on segmented cell.
   * Bring slider cell closer to Cocoa implementation.
   * Add ivar for Core Animation in NSView.
   * Improve border calculation on printing.
   * Lazy load app icon.
   * Better detection of removable volumes.
   * Polish translations.
   * Japanese translations.
   * Lots of bug fixes.

1.3 Noteworthy changes in version '0.26.2'
==========================================

This version is a small, but important bugfix release.

   * printing: Fix allocation of the CUPS printing classes.
   * installation: Fix the configure script.

1.4 Noteworthy changes in version '0.26.1'
==========================================

This version is released to conincide with version 1.25.1 of
gnustep-base, which contains changes required for this version of
gnustep-gui and gnustep-back.

   It includes an important workaround for users of GNUstep Objective-C
Runtime (libobjc2) and non-fragile ABI to avoid a bug in interaction
between the clang compiler and the runtime when non-fragile ABI is in
use.  Specifically, Clang and the runtime may disagree on what is the
offset of an ivar in a class's RAM. This manifested in a crash at
application startup due to misalignment of _gcontext inside NSThread.
See the mailing list discussion
(http://lists.gnu.org/archive/html/discuss-gnustep/2017-12/msg00129.html)
for more information.

   It also contains the following changes:

   * tests: Cleanup of warnings.
   * tests: Fix text system deallocation test.
   * printing: Undefine __BLOCKS__ before including cups.h, as some
     versions of the header expect that libdispatch is present and used
     if __BLOCKS__ is defined.
   * graphics context: Workaround for Clang+libobjc2+nonfragile ABI
     issue.

1.5 Noteworthy changes in version '0.26.0'
==========================================

This version was bumped due to previous binary incompatibilities between
0.25.0 and 0.25.1.  This version also includes numerous compatibility
improvements from the Summer of Code project, and a wide variety of
other fixes.  Notably, it fixes the use of cupsGetPPD() in the printing
system.

   * printing: Add an include to get deprecated function cupsGetPPD() on
     newer CUPS systems.
   * chore: Bump required base version.
   * tiff: Support for writing resolution.
   * jpeg: Save resolution information if it is different from 72 dpi.
   * save panel: Fix return type of sorting function.
   * events: Add some newer Cocoa enums and one method with dummy
     implementation.
   * speech synthesis: NSSpeechSynthesizerDelegate is now a @protocol on
     runtimes that support it.
   * pasteboard: New type identifiers.
   * translations: Some work on Polish, Russian and German translations
   * cell: Improvements to mouse tracking logic on NSCell.
   * image: If an unknown named image is unarchived with a coder or
     keyed coder, keep the name.
   * screen: Add -backingScaleFactor and return 1.0.
   * window: Return 1.0 from -backingScaleFactor.

   * compatibility: Numerous stub implementations of constants, classes
     and methods to improve source-level compatibility.
   * other bugfixes

1.6 Noteworthy changes in version '0.25.1'
==========================================

   * JPEG (saving) alpha channel fixes and size with resolution != 72
   * JPEG resolution read support
   * TIFF saving fixes
   * Improved volumes mounting and support
   * Portability improvements in volume mounting and support
   * Corrected layout of empty strings
   * Only update visible menus

1.7 Noteworthy changes in version '0.25.0'

gnu/usr.bin/avahi/avahi-utils/avahi-publish.c
/***
  This file is part of avahi.

  avahi is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) any later version.

  avahi is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General
  Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with avahi; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdlib.h>
#include <stdio.h>
#include <getopt.h>
#include <assert.h>
#include <string.h>
#include <sys/types.h>
#include <errno.h>
#include <locale.h>

#include <avahi-common/simple-watch.h>
#include <avahi-common/error.h>
#include <avahi-common/malloc.h>
#include <avahi-common/alternative.h>
#include <avahi-common/i18n.h>
#include <avahi-client/client.h>
#include <avahi-client/publish.h>

#include "sigint.h"

typedef enum {
    COMMAND_UNSPEC,
    COMMAND_HELP,
    COMMAND_VERSION,
    COMMAND_PUBLISH_SERVICE,
    COMMAND_PUBLISH_ADDRESS
} Command;

typedef struct Config {
    int verbose, no_fail, no_reverse;
    Command command;
    char *name, *stype, *domain, *host;
    uint16_t port;
    AvahiStringList *txt, *subtypes;
    AvahiAddress address;
} Config;

static AvahiSimplePoll *simple_poll = NULL;
static AvahiClient *client = NULL;
static AvahiEntryGroup *entry_group = NULL;

static int register_stuff(Config *config);

static void entry_group_callback(AvahiEntryGroup *g, AvahiEntryGroupState state, void *userdata) {
    Config *config = userdata;

    assert(g);
    assert(config);

    switch (state) {

        case AVAHI_ENTRY_GROUP_ESTABLISHED:

            fprintf(stderr, _("Established under name '%s'\n"), config->name);
            break;

        case AVAHI_ENTRY_GROUP_FAILURE:

            fprintf(stderr, _("Failed to register: %s\n"), avahi_strerror(avahi_client_errno(client)));
            break;

        case AVAHI_ENTRY_GROUP_COLLISION: {
            char *n;

            if (config->command == COMMAND_PUBLISH_SERVICE)
                n = avahi_alternative_service_name(config->name);
            else {
                assert(config->command == COMMAND_PUBLISH_ADDRESS);
                n = avahi_alternative_host_name(config->name);
            }

            fprintf(stderr, _("Name collision, picking new name '%s'.\n"), n);
            avahi_free(config->name);
            config->name = n;

            register_stuff(config);

            break;
        }

        case AVAHI_ENTRY_GROUP_UNCOMMITED:
        case AVAHI_ENTRY_GROUP_REGISTERING:
            ;
    }
}

static int register_stuff(Config *config) {
    assert(config);

    if (!entry_group) {
        if (!(entry_group = avahi_entry_group_new(client, entry_group_callback, config))) {
            fprintf(stderr, _("Failed to create entry group: %s\n"), avahi_strerror(avahi_client_errno(client)));
            return -1;
        }
    }

    assert(avahi_entry_group_is_empty(entry_group));

    if (config->command == COMMAND_PUBLISH_ADDRESS) {

        if (avahi_entry_group_add_address(entry_group, AVAHI_IF_UNSPEC, AVAHI_PROTO_UNSPEC, config->no_reverse ? AVAHI_PUBLISH_NO_REVERSE : 0, config->name, &config->address) < 0) {
            fprintf(stderr, _("Failed to add address: %s\n"), avahi_strerror(avahi_client_errno(client)));
            return -1;
        }

    } else {
        AvahiStringList *i;

        assert(config->command == COMMAND_PUBLISH_SERVICE);

        if (avahi_entry_group_add_service_strlst(entry_group, AVAHI_IF_UNSPEC, AVAHI_PROTO_UNSPEC, 0, config->name, config->stype, config->domain, config->host, config->port, config->txt) < 0) {
            fprintf(stderr, _("Failed to add service: %s\n"), avahi_strerror(avahi_client_errno(client)));
            return -1;
        }

        for (i = config->subtypes; i; i = i->next)
            if (avahi_entry_group_add_service_subtype(entry_group, AVAHI_IF_UNSPEC, AVAHI_PROTO_UNSPEC, 0, config->name, config->stype, config->domain, (char*) i->text) < 0) {
                fprintf(stderr, _("Failed to add subtype '%s': %s\n"), i->text, avahi_strerror(avahi_client_errno(client)));
                return -1;
            }
    }

    avahi_entry_group_commit(entry_group);

    return 0;
}

static void client_callback(AvahiClient *c, AvahiClientState state, AVAHI_GCC_UNUSED void * userdata) {
    Config *config = userdata;

    client = c;

    switch (state) {
        case AVAHI_CLIENT_FAILURE:

            if (config->no_fail && avahi_client_errno(c) == AVAHI_ERR_DISCONNECTED) {
                int error;

                /* We have been disconnected, so let reconnect */

                fprintf(stderr, _("Disconnected, reconnecting ...\n"));

                avahi_client_free(client);
                client = NULL;
                entry_group = NULL;

                if (!(client = avahi_client_new(avahi_simple_poll_get(simple_poll), AVAHI_CLIENT_NO_FAIL, client_callback, config, &error))) {
                    fprintf(stderr, _("Failed to create client object: %s\n"), avahi_strerror(error));
                    avahi_simple_poll_quit(simple_poll);
                }

            } else {
                fprintf(stderr, _("Client failure, exiting: %s\n"), avahi_strerror(avahi_client_errno(c)));
                avahi_simple_poll_quit(simple_poll);
            }

            break;

        case AVAHI_CLIENT_S_RUNNING:

            if (register_stuff(config) < 0)
                avahi_simple_poll_quit(simple_poll);

            break;

        case AVAHI_CLIENT_S_COLLISION:

            if (config->verbose)
                fprintf(stderr, _("Host name conflict\n"));

            /* Fall through */

        case AVAHI_CLIENT_S_REGISTERING:

            if (entry_group) {
                avahi_entry_group_free(entry_group);
                entry_group = NULL;
            }
            break;

        case AVAHI_CLIENT_CONNECTING:

            if (config->verbose)
                fprintf(stderr, _("Waiting for daemon ...\n"));

            break;

            ;
    }
}

static void help(FILE *f, const char *argv0) {
    fprintf(f,
            _("%s [options] %s <name> <type> <port> [<txt ...>]\n"
              "%s [options] %s <host-name> <address>\n\n"
              "    -h --help            Show this help\n"
              "    -V --version         Show version\n"
              "    -s --service         Publish service\n"
              "    -a --address         Publish address\n"
              "    -v --verbose         Enable verbose mode\n"
              "    -d --domain=DOMAIN   Domain to publish service in\n"
              "    -H --host=DOMAIN     Host where service resides\n"
              "       --subtype=SUBTYPE An additional subtype to register this service with\n"
              "    -R --no-reverse      Do not publish reverse entry with address\n"
              "    -f --no-fail         Don't fail if the daemon is not available\n"),
              argv0, strstr(argv0, "service") ? "[-s]" : "-s",
              argv0, strstr(argv0, "address") ? "[-a]" : "-a");
}

static int parse_command_line(Config *c, const char *argv0, int argc, char *argv[]) {
    int o;

    enum {
        ARG_SUBTYPE = 256
    };

    static const struct option long_options[] = {
        { "help",           no_argument,       NULL, 'h' },
        { "version",        no_argument,       NULL, 'V' },
        { "service",        no_argument,       NULL, 's' },
        { "address",        no_argument,       NULL, 'a' },
        { "verbose",        no_argument,       NULL, 'v' },
        { "domain",         required_argument, NULL, 'd' },
        { "host",           required_argument, NULL, 'H' },
        { "subtype",        required_argument, NULL, ARG_SUBTYPE},
        { "no-reverse",     no_argument,       NULL, 'R' },
        { "no-fail",        no_argument,       NULL, 'f' },
        { NULL, 0, NULL, 0 }
    };

    assert(c);

    c->command = strstr(argv0, "address") ? COMMAND_PUBLISH_ADDRESS : (strstr(argv0, "service") ? COMMAND_PUBLISH_SERVICE : COMMAND_UNSPEC);
    c->verbose = c->no_fail = c->no_reverse = 0;
    c->host = c->name = c->domain = c->stype = NULL;
    c->port = 0;
    c->txt = c->subtypes = NULL;

    while ((o = getopt_long(argc, argv, "hVsavRd:H:f", long_options, NULL)) >= 0) {

        switch(o) {
            case 'h':
                c->command = COMMAND_HELP;
                break;
            case 'V':
                c->command = COMMAND_VERSION;
                break;
            case 's':
                c->command = COMMAND_PUBLISH_SERVICE;
                break;
            case 'a':
                c->command = COMMAND_PUBLISH_ADDRESS;
                break;
            case 'v':
                c->verbose = 1;
                break;
            case 'R':
                c->no_reverse = 1;
                break;
            case 'd':
                avahi_free(c->domain);
                c->domain = avahi_strdup(optarg);
                break;
            case 'H':
                avahi_free(c->host);
                c->host = avahi_strdup(optarg);
                break;
            case 'f':
                c->no_fail = 1;
                break;
            case ARG_SUBTYPE:
                c->subtypes = avahi_string_list_add(c->subtypes, optarg);
                break;
            default:
                return -1;
        }
    }

    if (c->command == COMMAND_PUBLISH_ADDRESS) {
        if (optind+2 !=  argc) {
            fprintf(stderr, _("Bad number of arguments\n"));
            return -1;
        }

        avahi_free(c->name);
        c->name = avahi_strdup(argv[optind]);
        avahi_address_parse(argv[optind+1], AVAHI_PROTO_UNSPEC, &c->address);

    } else if (c->command == COMMAND_PUBLISH_SERVICE) {

        char *e;
        long int p;
        int i;

        if (optind+3 > argc) {
            fprintf(stderr, _("Bad number of arguments\n"));
            return -1;
        }

        c->name = avahi_strdup(argv[optind]);
        c->stype = avahi_strdup(argv[optind+1]);

        errno = 0;
        p = strtol(argv[optind+2], &e, 0);

        if (errno != 0 || *e || p < 0 || p > 0xFFFF) {
            fprintf(stderr, _("Failed to parse port number: %s\n"), argv[optind+2]);
            return -1;
        }

        c->port = p;

        for (i = optind+3; i < argc; i++)
            c->txt = avahi_string_list_add(c->txt, argv[i]);
    }

    return 0;
}

int main(int argc, char *argv[]) {
    int ret = 1, error;
    Config config;
    const char *argv0;

    avahi_init_i18n();
    setlocale(LC_ALL, "");

    if ((argv0 = strrchr(argv[0], '/')))
        argv0++;
    else
        argv0 = argv[0];

    if (parse_command_line(&config, argv0, argc, argv) < 0)
        goto fail;

    switch (config.command) {
        case COMMAND_UNSPEC:
            ret = 1;
            fprintf(stderr, _("No command specified.\n"));
            break;

        case COMMAND_HELP:
            help(stdout, argv0);
            ret = 0;
            break;

        case COMMAND_VERSION:
            printf("%s "PACKAGE_VERSION"\n", argv0);
            ret = 0;
            break;

        case COMMAND_PUBLISH_SERVICE:
        case COMMAND_PUBLISH_ADDRESS:

            if (!(simple_poll = avahi_simple_poll_new())) {
                fprintf(stderr, _("Failed to create simple poll object.\n"));
                goto fail;
            }

            if (sigint_install(simple_poll) < 0)
                goto fail;

            if (!(client = avahi_client_new(avahi_simple_poll_get(simple_poll), config.no_fail ? AVAHI_CLIENT_NO_FAIL : 0, client_callback, &config, &error))) {
                fprintf(stderr, _("Failed to create client object: %s\n"), avahi_strerror(error));
                goto fail;
            }

            if (avahi_client_get_state(client) != AVAHI_CLIENT_CONNECTING && config.verbose) {
                const char *version, *hn;

                if (!(version = avahi_client_get_version_string(client))) {
                    fprintf(stderr, _("Failed to query version string: %s\n"), avahi_strerror(avahi_client_errno(client)));
                    goto fail;
                }

                if (!(hn = avahi_client_get_host_name_fqdn(client))) {
                    fprintf(stderr, _("Failed to query host name: %s\n"), avahi_strerror(avahi_client_errno(client)));
                    goto fail;
                }

                fprintf(stderr, _("Server version: %s; Host name: %s\n"), version, hn);
            }

            avahi_simple_poll_loop(simple_poll);
            ret = 0;
            break;
    }

fail:

    if (client)
        avahi_client_free(client);

    sigint_uninstall();

    if (simple_poll)
        avahi_simple_poll_free(simple_poll);

    avahi_free(config.host);
    avahi_free(config.name);
    avahi_free(config.stype);
    avahi_free(config.domain);
    avahi_string_list_free(config.subtypes);
    avahi_string_list_free(config.txt);

    return ret;
}
